### PR TITLE
Add multi-target support

### PR DESCRIPTION
**Pull request by pincher95**
### PR: Add multi-target `/probe` endpoint + auth-modules (fixes [#1062](https://github.com/prometheus-community/elasticsearch_exporter/issues/1062))

This PR implements the feature described in issue [#1062](https://github.com/prometheus-community/elasticsearch_exporter/issues/1062) (“multi-target scraping”).

───────────────────────────────
What’s new
───────────────────────────────
1. **`/probe` endpoint**  
   Accepts `?target=` and (optionally) `auth_module=`.  
   For each request the exporter spins up a throw-away Prometheus registry, registers the usual collectors with the supplied ES URL, gathers, and streams metrics.

2. **Auth-module support (`--config.file=`)**  
   Simple YAML schema modelled after `postgres_exporter` & `mysqld_exporter`.

   ```yaml
   auth_modules:
     prod_basic:
       type: userpass
       userpass:
         username: metrics
         password: s3cr3t

     staging_key:
       type: apikey
       apikey: BASE64IDKEY==

     # options: <k:v> are appended as DSN query parameters when present
   ```

3. **Alias on `/metrics`**  
   Requests that already contain `target=` are internally forwarded to the probe handler, so users may keep `path: /metrics` in ServiceMonitors if they wish.

4. **Per-module API key support**  
   `type: apikey` injects the `Authorization: ApiKey …` header for that probe only.

5. **Docs & examples**  
   * README: updated multi-target section with full table of supported auth types.  
   * `examples/auth_modules.yml` shows `userpass`, `apikey`.

6. **Unit tests**  
   * `config` loader round-trip.  
   * Validation edge cases for `validateProbeParams`.

───────────────────────────────
Compatibility
───────────────────────────────
* **Single-target usage unchanged** – `/metrics` without `target=` still returns the startup cluster.
* All existing CLI flags (TLS, indices, AWS signing, …) still apply.
* No exported metrics renamed.

───────────────────────────────
Prometheus sample config
───────────────────────────────
```yaml
- job_name: es-multi
  metrics_path: /probe
  params:
    auth_module: [prod_basic]
  static_configs:
    - targets:
        - https://es-prod:9200
    - targets:
        - https://es-staging:9200
      labels:
        __param_auth_module: staging_key
  relabel_configs:
    - source_labels: [__address__]
      target_label: __param_target
    - source_labels: [__param_target]
      target_label: instance
    - target_label: __address__
      replacement: exporter:9114
```

───────────────────────────────
Commit outline
───────────────────────────────
1. refactor: probe handler, alias logic  
2. feat: YAML auth-module loader (userpass, apikey)  
3. docs: README + example file  
4. tests: config loader & validation  
5. CHANGELOG entry

Feedback / suggestions welcome!

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package config
+
+import (
+	"os"
+
+	"gopkg.in/yaml.v3"
```
Comment (sysadmind) [inline config/config.go:6]: This package is deprecated and the project is archived. This is the continuation and should be a drop in replacement -> https://github.com/yaml/go-yaml

I know that postgres_exporter uses this package and this is copy of that, but I hate to bring in a new dependency that is already deprecated.

File: README.md [hunk 2]
```diff
@@ -95,211 +96,264 @@ Specifying those two environment variables will override authentication passed i
...
+| elasticsearch_health_report_restarting_primaries                     | gauge   | 1           | The number of restarting primary shards                                                             |
+| elasticsearch_health_report_restarting_replicas                      | gauge   | 1           | The number of restarting replica shards                                                             |
+| elasticsearch_health_report_shards_availabilty_status                | gauge   | 2           | Shards availabilty status                                                                           |
+| elasticsearch_health_report_shards_capacity_status                   | gauge   | 2           | Shards capacity status                                                                              |
+| elasticsearch_health_report_slm_policies                             | gauge   | 1           | The number of SLM policies                                                                          |
+| elasticsearch_health_report_slm_status                               | gauge   | 2           | SLM status                                                                                          |
```
Comment (sysadmind) [inline README.md:350]: Can you please pull the whitespace changes out? They could be in a different PR but they're unrelated here and make it harder to see the actual changes.

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,10 @@ The flag `--es.ilm` has been renamed to `--collector.ilm`.
 * [CHANGE] Rename --es.data_stream to --collector.data-stream #983
 * [CHANGE] Rename --es.ilm to --collector.ilm #999
 
+## [Unreleased]
```
Comment (sysadmind) [inline CHANGELOG.md:11]: There is already an unreleased section above this. Please use that section, following the format that already exists.

File: main.go [hunk 4]
```diff
@@ -60,7 +61,7 @@ func main() {
 		toolkitFlags = webflag.AddFlags(kingpin.CommandLine, ":9114")
 		esURI        = kingpin.Flag("es.uri",
 			"HTTP API address of an Elasticsearch node.").
-			Default("http://localhost:9200").String()
+			Default("").String()
```
Comment (sysadmind) [inline main.go:64]: I think that this would be a breaking change that would need to be called out.

**Review by sysadmind — CHANGES_REQUESTED** [review]
I need to spend more time with the code to provide all of my feedback, but want to provide what I have ready while I gave it the first pass.

File: README.md (hunk 2 above)
Comment (pincher95) [reply README.md]: Done

File: config/config_test.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package config
+
+import (
+	"os"
+	"testing"
+)
+
+func TestLoadConfig(t *testing.T) {
```
Comment (sysadmind) [inline config/config_test.go:8]: I think that it would be good to have at least one test for the apikey type as well. Technically you could set both configs, but the type should determine what is used in the code.

File: examples/auth_modules.yml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      sslmode: disable # example option – becomes ?sslmode=disable
+
+  # 2. Read-only account for staging cluster
+  staging_ro:
+    type: userpass
+    userpass:
+      username: readonly
+      password: changeme
+
+  # 3. API-Key authentication (future)
```
Comment (sysadmind) [inline examples/auth_modules.yml:22]: I don't think that this is future. This PR includes this type.

File: main.go [hunk 7]
```diff
@@ -129,116 +143,132 @@ func main() {
...
-		os.Exit(1)
+		// register cluster info retriever as prometheus collector
+		prometheus.MustRegister(clusterInfoRetriever)
 	}
 
-	// register cluster info retriever as prometheus collector
-	prometheus.MustRegister(clusterInfoRetriever)
+	http.HandleFunc(*metricsPath, func(w http.ResponseWriter, r *http.Request) {
+		// If query has target param treat like probe endpoint
+		if r.URL.Query().Has("target") {
```
Comment (sysadmind) [inline main.go:261]: I'm not sure that I see the value in having the ability to swap to multi-target support here. The /probe endpoint could do things that the /metrics endpoint would not which would be unexpected for users. I don't know of any exporters that handle the target parameter like this so I don't know that it's a good idea to do so here.

File: probe.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	am, ok := cfg.AuthModules[modu]
+	if !ok {
+		return "", nil, errModuleNotFound
+	}
+	switch strings.ToLower(am.Type) {
+	case "userpass":
+		if am.UserPass != nil {
+			return target, &am, nil
+		}
+		return "", nil, errUnsupportedModule
```
Comment (sysadmind) [inline probe.go:52]: This error appears to be inaccurate. This really means that the type was set to userpass, but username and password weren't actually set. This would be better handled in the config load and validation in my opinion.

File: probe.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+	case "apikey":
+		if am.APIKey != "" {
+			return target, &am, nil
+		}
+		return "", nil, errUnsupportedModule
```
Comment (sysadmind) [inline probe.go:57]: ditto

**Review by sysadmind — CHANGES_REQUESTED** [review]
I think the guts of this are really solid. It did take me some time to give a proper review so once the changes are pushed up, I will need to allocate some time to re-review. Thanks for submitting this, I think it will be a good improvement.

File: examples/auth_modules.yml (hunk 6 above)
Comment (pincher95) [reply examples/auth_modules.yml]: removed "future" remark

File: config/config_test.go (hunk 5 above)
Comment (pincher95) [reply config/config_test.go:8]: added test for apikey

File: main.go [hunk 10]
```diff
@@ -264,9 +294,83 @@ func main() {
...
+			transport = &transportWithAPIKey{
+				underlyingTransport: transport,
+				apiKey:              am.APIKey,
+			}
+		}
+		probeClient := &http.Client{
+			Timeout:   *esTimeout,
+			Transport: transport,
+		}
+
```
[the remark this answers is not in the document]
Comment (pincher95) [reply main.go:347]: Added AWS auth support and AWS auth validation

File: probe.go (hunk 8 above)
Comment (pincher95) [reply probe.go]: Load time validation moved to the config package, will validate the auth configuration during start-up

File: probe.go (hunk 9 above)
Comment (pincher95) [reply probe.go:56]: Same as before

File: main.go [hunk 11]
```diff
@@ -264,9 +294,83 @@ func main() {
...
+				q := targetURL.Query()
+				for k, v := range am.Options {
+					q.Set(k, v)
+				}
+				targetURL.RawQuery = q.Encode()
+			}
+		}
+
+		// Build a dedicated HTTP client for this probe request (reuse TLS opts, timeout, etc.).
+		tlsCfg := createTLSConfig(*esCA, *esClientCert, *esClientPrivateKey, *esInsecureSkipVerify)
```
[the remark this answers is not in the document]
Comment (pincher95) [reply main.go]: 1. added tls to config per target
2. added tls config validation
3. updated config test to incluse tls config

**Review by sysadmind — CHANGES_REQUESTED** [review]
All of the new files are also missing the copyright assignment/license header. You should be able to copy paste (make sure there are no dates) from any other go file. I don't love the extra requests for the two collectors but I think that can be cleaned up after #1052 is merged.

File: collector/indices.go [hunk 12]
```diff
@@ -620,13 +620,28 @@ func (i *Indices) fetchAndDecodeIndexStats(ctx context.Context) (indexStatsRespo
...
+// getClusterName returns the cluster name. If no clusterinfo retriever is
+// attached (e.g. /probe mode) it performs a lightweight call to the root
+// endpoint once and caches the result.
 func (i *Indices) getClusterName() string {
-	if i.lastClusterInfo != nil {
+	if i.lastClusterInfo != nil && i.lastClusterInfo.ClusterName != "unknown_cluster" {
 		return i.lastClusterInfo.ClusterName
 	}
+	u := *i.url
+	u.Path = "/"
```
[the remark this answers is not in the document]
Comment (pincher95) [reply collector/indices.go]: fixed relative URLs

File: main.go [hunk 13]
```diff
@@ -264,9 +286,105 @@ func main() {
...
+				targetURL.RawQuery = q.Encode()
+			}
+		}
+
+		// Build a dedicated HTTP client for this probe request (reuse TLS opts, timeout, etc.).
+		pemCA := *esCA
+		pemCert := *esClientCert
+		pemKey := *esClientPrivateKey
+		insecure := *esInsecureSkipVerify
+		if am != nil && am.TLS != nil {
```
[the remark this answers is not in the document]
Comment (pincher95) [reply main.go:320]: Added tls as it own auth type
What’s allowed
1. userpass + tls (TLS for transport, basic-auth for identity)
2. apikey + tls (TLS for transport, API key for identity)
3. aws + tls (TLS for transport, SigV4 for identity)
4. tls only (client-cert auth; no userpass/apikey/aws)

File: config/config.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+			}
+		default:
+			return fmt.Errorf("auth_module %s has unsupported type %s", name, am.Type)
+		}
+
+		// TLS validation (optional but if specified must be coherent)
+		if am.TLS != nil {
+			if (am.TLS.CertFile == "") != (am.TLS.KeyFile == "") {
+				return fmt.Errorf("auth_module %s tls requires both cert_file and key_file or neither", name)
+			}
```
[the remark this answers is not in the document]
Comment (pincher95) [reply config/config.go]: Added TLS as a distinct authentication type

File: config/config.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+		case "userpass":
+			if am.UserPass == nil || am.UserPass.Username == "" || am.UserPass.Password == "" {
+				return fmt.Errorf("auth_module %s type userpass requires username and password", name)
+			}
+		case "apikey":
+			if am.APIKey == "" {
+				return fmt.Errorf("auth_module %s type apikey requires apikey", name)
+			}
+		case "aws":
+			if am.AWS == nil || am.AWS.Region == "" {
```
[the remark this answers is not in the document]
Comment (pincher95) [reply config/config.go:71]: I disagree, AWS signing requires a region to and it's not loaded from the environment as far as i can see

[the remark this answers is not in the document]
Comment (sysadmind) [reply config/config.go:71]: The AWS function `LoadDefaultConfig` will parse environment variables and look for the config files in well known locations. It will eventually make its way here: https://github.com/aws/aws-sdk-go-v2/blob/main/config/defaultsmode.go#L28

File: config/config.go [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+			return fmt.Errorf("auth_module %s has unsupported type %s", name, am.Type)
+		}
+
+		// Validate TLS configuration (optional for all auth types, provides transport security)
+		if am.TLS != nil {
+			// For cert-based auth (type: tls), cert and key are required
+			// For other auth types, TLS config is optional and used for transport security
+			if strings.ToLower(am.Type) == "tls" {
+				// Already validated above that cert and key are present
+			} else {
```
Comment (sysadmind) [inline config/config.go:102]: This could be simplified
```suggestion
			if !strings.ToLower(am.Type) == "tls" {
```

File: config/config_test.go [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    tls:
+      ca_file: ` + ca + `
+      cert_file: ` + cert + `
+      key_file: ` + key + ``,
+	}}
+
+	for _, c := range positive {
+		tmp, _ := os.CreateTemp(t.TempDir(), "cfg-*.yml")
+		tmp.WriteString(c.yaml)
+		tmp.Close()
```
Comment (sysadmind) [inline config/config_test.go:84]: These files don't ever seem to be cleaned up.

File: config/config_test.go [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+	}
+	f.Close()
+	return f.Name()
+}
+
+// ---------------------------- Positive cases ----------------------------
+func TestLoadConfigPositiveVariants(t *testing.T) {
+	ca := mustTempFile(t)
+	cert := mustTempFile(t)
+	key := mustTempFile(t)
```
Comment (sysadmind) [inline config/config_test.go:21]: These files don't ever seem to be cleaned up.

**Review by sysadmind — CHANGES_REQUESTED** [review]
The DCO check is also failing on at least one commit

File: main.go [hunk 19]
```diff
@@ -264,9 +286,105 @@ func main() {
...
+		}
+
+		// inject authentication based on auth_module type
+		if am != nil {
+			if strings.EqualFold(am.Type, "apikey") && am.APIKey != "" {
+				transport = &transportWithAPIKey{
+					underlyingTransport: transport,
+					apiKey:              am.APIKey,
+				}
+			} else if strings.EqualFold(am.Type, "aws") && am.AWS != nil {
```
[the remark this answers is not in the document]
Comment (pincher95) [reply main.go]: region validation removed

File: config/config.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (pincher95) [reply config/config.go]: region validation removed

File: config/config_test.go (hunk 18 above)
Comment (pincher95) [reply config/config_test.go:24]: added `t.Cleanup` to remove tmp files

File: main.go [hunk 20]
```diff
@@ -129,116 +143,124 @@ func main() {
 	promslogConfig.Writer = w
 	logger := promslog.New(promslogConfig)
 
-	esURL, err := url.Parse(*esURI)
-	if err != nil {
-		logger.Error("failed to parse es.uri", "err", err)
-		os.Exit(1)
-	}
+	// version metric
+	prometheus.MustRegister(versioncollector.NewCollector(name))
```
[the remark this answers is not in the document]
Comment (pincher95) [reply main.go:147]: added metric to `/probe`

File: config/config.go [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+			if am.UserPass == nil || am.UserPass.Username == "" || am.UserPass.Password == "" {
+				return fmt.Errorf("auth_module %s type userpass requires username and password", name)
+			}
+		case "apikey":
+			if am.APIKey == "" {
+				return fmt.Errorf("auth_module %s type apikey requires apikey", name)
+			}
+		case "aws":
+			if am.AWS == nil {
+				return fmt.Errorf("auth_module %s type aws requires region", name)
```
[the remark this answers is not in the document]
Comment (pincher95) [reply config/config.go]: Removed strict validation for AWS region and and removed wrong error

**Review by sysadmind — APPROVED** [review]

**Comment by sysadmind** [comment]
Thanks! Great stuff! I appreciate all of your hard work on this one to get it ready to ship.
